# Calculation of RSE and quantitative determination of PFAS concentrations from the LC-TQ
Author: Valerie de Rijk

Date: 15/8/2025

## 1 Data Loading
First we load in the raw data as a csv file (which is the file you directly get from MassHunter).

In [1]:
%run src/functions_quantitative.ipynb
rawdata = load_and_label_pfas_csv("rawdata/20250828_AllResults_v4.csv")

Detected delimiter: ','
['PFBA', '13C4-PFBA', 'PFPeA', '13C5-PFPeA', 'HFPO-DA', '13C3-HFPO-DA', 'PFBS', '13C3-PFBS', 'PFHxA', '13C5-PFHxA', '4:2 FTS', '13C2-4:2FTS', 'PFPeS', 'PFPHpA', '13C4-PFHpA', 'ADONA', 'PFHxS', '13C3-PFHxS', 'PFOA', '13C8-PFOA', '6:2 FTS', '13C2-6:2FTS', 'PFHpS', 'PFNA', '13C9-PFNA', 'PFOSA', 'PFOS', '13C8-PFOSA', '13C8-PFOS', 'NMeFOSA', 'PFDA', 'D3-NMeFOSA', '13C6-PFDA', 'NEtFOSA', '8:2 FTS', '13C2-8:2FTS', '9Cl-PF3ONS', 'D5-NEtFOSA', 'PFNS', 'PFUnA', 'NMeFOSAA', '13C7-PFUnA', 'D3-NMeFOSAA', 'NEtFOSAA', 'D5-NEtFOSAA', 'PFDS', 'PFDoA', '13C2-PFDoA', 'NMeFOSE', 'D7-NMeFOSE', 'NEtFOSE', '11Cl-PF3OUdS', 'D9-NEtFOSE', 'PFTrDA', 'PFTeDA', '13C2-PFTeDA']
Total compounds: 56
Without isotope prefix: 32


## 2  Calibration Line (Instrument linearity)
To analyze all instrument linearity metrics, we first need to calculate the concentrations found in the samples. 
We do this by caclulating the concentrations based on the formula on page 44 of EPA method 1633. 
To achieve this, we first need to calculate the average response factor per compound. 


We compute the RSE based on  the formula for RSE based on EPA method 1633 (section 10.3.3.3, page 28, option 2). The functions are loaded in a different script. 
If the RSE remains above 20%, this means something is wrong with the linearity of the LC-MS TQ and the calculated concentrations in section 3 are not trustworthy. 

*We mute the output of this cell for clarity of the script and summarize them under EIS results and Target Analytes results. If interested in full computation, please unmute the cell by removing the cell line **%%capture***

The output is structured as following: 
- List of identified EIS compounds (check if all are found)
- Computation for the EIS compounds
- Computation for the target analytes
- Summary of the result for EIS
- Summary of the result for target

*note: if not interested in detailed calculations, run the cell and ignore the output*


In [2]:
%%capture
%run src/functions_quantitative.ipynb
import importlib
import expected_concs
from expected_concs import expected_concs_EIS, expected_concs_target_analytes

from analogs_compounds import target_EIS_analogs, EIS_NIS_analogs
from solutions import calibration_solutions

# computation of EIS concentration and recovery rate
from solutions import NIS_stock, EIS_stock
from allowed_recoveries import EIS_NIS_compound_recovery_aqueous

eis_RFS = calculate_average_RFs_EIS(rawdata, EIS_NIS_analogs, calibration_solutions)


Ws_cal = 0.25#ml
dilution_NIS_stock_cal = 40 #dilution cal
dilution_EIS_stock_cal = 40 #dilution cal
spiked_amount_cal = 0.05 #ml
Df_cal = 1 #dilution factor
spiked_amount = 0.05 #ml

dilution_EIS_stock = 2 
spiked_amount_EIS =0.05 #ml

#calibration line
conc_EIS_cal = calculate_conc_EIS_cal(rawdata, eis_RFS, dilution_NIS_stock_cal, Ws_cal,
                          EIS_NIS_analogs, Df_cal, NIS_stock, spiked_amount_cal)

calculated_recoveries_EIS_cal = calculate_recoveries_EIS(conc_EIS_cal, EIS_stock, dilution_EIS_stock_cal, spiked_amount_cal, Ws_cal)

rse_results_cal_EIS = calculate_eis_rse_modified(conc_EIS_cal, expected_concs_EIS, p=2)
validation_results_cal= validate_recoveries(calculated_recoveries_EIS_cal, EIS_NIS_compound_recovery_aqueous)
summary_cal = summarize_validation_results(validation_results_cal)
print("Validation Summary:")
for key, value in summary_cal['Overall'].items():
    print(f"{key}: {value}")

failed_recoveries_cal = get_failed_recoveries(validation_results_cal)



# Example run
#rse_results_eis = calculate_eis_rse(rawdata, expected_concs_EIS)
#rse_results_target = calculate_target_rse(rawdata, expected_concs_target_analytes)

#print_rse_summary(rse_results_eis)
#print_rse_summary(rse_results_target)

TypeError: calculate_recoveries_EIS() takes 2 positional arguments but 5 were given

### 2.1 EIS results

We print the results in a clear table and save them to csv (first for EIS, then target). Give the files an intuitive name.

In [3]:
df_rse_eis=df_rse_results(rse_results_cal_EIS)
display(df_rse_eis)

Compound  RSE_Percent  ...  Pass_Fail Sum_Relative_Squared_Error
0     13C2-PFDoA     8.323322  ...       PASS                   0.006928
1      13C8-PFOA     3.154659  ...       PASS                   0.000995
2    13C2-6:2FTS    10.675406  ...       PASS                   0.011396
3      13C9-PFNA     2.895678  ...       PASS                   0.000838
4     13C5-PFHxA     0.909806  ...       PASS                   0.000083
5      13C4-PFBA     0.383990  ...       PASS                   0.000015
6     13C3-PFHxS     2.561264  ...       PASS                   0.000656
7   13C3-HFPO-DA     1.599224  ...       PASS                   0.000256
8      13C3-PFBS     2.989945  ...       PASS                   0.000894
9    13C2-PFTeDA     7.067921  ...       PASS                   0.004996
10   D3-NMeFOSAA     5.882618  ...       PASS                   0.003461
11    13C4-PFHpA     4.955673  ...       PASS                   0.002456
12   13C2-8:2FTS    24.611513  ...       FAIL                   0.060573
13     13C6-PFDA     6.361348  ...       PASS                   0.004047
14    13C8-PFOSA     5.697557  ...       PASS                   0.003246
15    D9-NEtFOSE    15.059593  ...       PASS                   0.022679
16    D5-NEtFOSA    12.419995  ...       PASS                   0.015426
17    13C5-PFPeA     3.118019  ...       PASS                   0.000972
18   13C2-4:2FTS     6.157341  ...       PASS                   0.003791
19    D7-NMeFOSE    14.146098  ...       PASS                   0.020011
20    13C7-PFUnA     8.834679  ...       PASS                   0.007805
21   D5-NEtFOSAA     5.109575  ...       PASS                   0.002611
22    D3-NMeFOSA    14.120633  ...       PASS                   0.019939
23     13C8-PFOS     6.436924  ...       PASS                   0.004143

[24 rows x 7 columns]

### 2.2 Target Analytes results

In [6]:
df_rse_target = df_rse_results(rse_results_target)
display(df_rse_target)

Compound  RSE_Percent  ...  Pass_Fail Sum_Relative_Squared_Error
0           PFBA    13.126824  ...       PASS                   0.017231
1          PFPeA    12.245095  ...       PASS                   0.014994
2        HFPO-DA    11.736307  ...       PASS                   0.013774
3           PFBS    12.797400  ...       PASS                   0.016377
4          PFHxA    10.595762  ...       PASS                   0.011227
5        4:2 FTS    27.750370  ...       FAIL                   0.077008
6          PFPeS    15.716212  ...       PASS                   0.024700
7          ADONA    14.804778  ...       PASS                   0.021918
8          PFHxS    16.833835  ...       PASS                   0.028338
9           PFOA    14.110585  ...       PASS                   0.019911
10       6:2 FTS    11.150538  ...       PASS                   0.012433
11         PFHpS    12.751320  ...       PASS                   0.016260
12          PFNA    11.386971  ...       PASS                   0.012966
13         PFOSA    12.424319  ...       PASS                   0.015436
14          PFOS    14.428503  ...       PASS                   0.020818
15       NMeFOSA    12.573516  ...       PASS                   0.015809
16          PFDA    13.157659  ...       PASS                   0.017312
17       NEtFOSA    23.221161  ...       FAIL                   0.053922
18       8:2 FTS    24.311037  ...       FAIL                   0.059103
19    9Cl-PF3ONS    15.979258  ...       PASS                   0.025534
20          PFNS    23.659415  ...       FAIL                   0.055977
21         PFUnA    12.576632  ...       PASS                   0.015817
22      NMeFOSAA    15.611611  ...       PASS                   0.024372
23      NEtFOSAA    18.173101  ...       PASS                   0.033026
24          PFDS    14.740593  ...       PASS                   0.021729
25         PFDoA    13.418838  ...       PASS                   0.018007
26       NMeFOSE   165.332210  ...       FAIL                   2.733474
27       NEtFOSE   174.257586  ...       FAIL                   3.036571
28  11Cl-PF3OUdS    17.293324  ...       PASS                   0.029906
29        PFTrDA    16.293633  ...       PASS                   0.026548
30        PFTeDA    12.645325  ...       PASS                   0.015990

[31 rows x 7 columns]

## 3 Instrument checks
*all of these calculations still need to be added*

We perform different checks apart from instrument linearity to ensure that the LC-TQ is behaving as we expect. 
1. Instrument Sensitivity check (signal to noise ratio)
     - here we check if the ratio is greater than or equal to 3:1 (EPA section 14.1)
2. Bile salt interference check
   - We check that the RT of bile salts falls at least one minute outside of the retention time for any linear or branched PFOS (EPA 14.2)
3. Calibration verification (EPA 14.3)
   - Check that the CV, e.g. the mid-level calibration point, (which should be done every 10 samples or less) has a recovery between 70 and 130%.
4. Retention times
    - All target analyte with  true isotope dilution should elute within 0.1 minutes of the associated EIS compounds.
5. Ongoing precision and recovery
    - The recovery of native compounds and EIS compounds must meet the standards in table 5,6,7 and 8 of the EPA.
6. Instrument blank and method blank
   - Check that is it below all analyte's MDL.
7. Instrument sensitivity
   - This is the ratio of the NIS peak areas from QC and field samples relative to the mean area of the corresponding NIS in most recent inital calibration. This must be within 50 to 200%. 


## 3.1 recovery  EIS compounds
This follows EPA method 1633 page 44 and EPA method page 28 (equation for RFs). We first compute the average response factor for the EIS compounds in the calibration standard. Then, we use the formula on page 44 to calculate the concentration in the sample and compare it to the expected concentration to determine the recovery. 

We also check the used concentration from above. 

### 3.1.1 Calculating average Response Factor EIS compounds

In [2]:
from analogs_compounds import target_EIS_analogs, EIS_NIS_analogs
from solutions import calibration_solutions

eis_RFS = calculate_average_RFs_EIS(rawdata, EIS_NIS_analogs, calibration_solutions)


AVERAGE RESPONSE FACTORS FOR EIS COMPOUNDS (n=24)


Compound_Name  NIS_Analog  ...  Average_Response_Factor   RSD_%
0     13C3-PFHxS  18O2-PFHxS  ...                   1.8312  173.43
1     13C5-PFHxA  13C2-PFHxA  ...                   2.0985  173.16
2     13C2-PFDoA   13C2-PFDA  ...                   3.8255  171.54
3     13C5-PFPeA  13C2-PFHxA  ...                   1.0163  173.06
4    13C2-6:2FTS  18O2-PFHxS  ...                   4.8377  163.79
5    D5-NEtFOSAA   13C4-PFOS  ...                   0.4679  170.42
6     D5-NEtFOSA   13C4-PFOS  ...                   8.8892  170.34
7     13C4-PFHpA  13C2-PFHxA  ...                   1.9642  172.48
8      13C9-PFNA   13C5-PFNA  ...                   4.0900  173.79
9   13C3-HFPO-DA  13C2-PFHxA  ...                   0.5044  171.98
10    D7-NMeFOSE   13C4-PFOS  ...                   0.1928  173.16
11    13C7-PFUnA   13C2-PFDA  ...                   0.9460  168.67
12   13C2-PFTeDA   13C2-PFDA  ...                   3.5936  168.84
13    D9-NEtFOSE   13C4-PFOS  ...                   0.5050  175.47
14   13C2-8:2FTS  18O2-PFHxS  ...                   4.1787  171.13
15     13C6-PFDA   13C2-PFDA  ...                   1.1411   11.93
16     13C4-PFBA   13C3-PFBA  ...                   0.9490  173.38
17     13C8-PFOS   13C4-PFOS  ...                   3.9500    4.82
18     13C8-PFOA   13C4-PFOA  ...                   1.8226  171.74
19   D3-NMeFOSAA   13C4-PFOS  ...                   0.8343    7.47
20     13C3-PFBS  18O2-PFHxS  ...                   1.8849  172.66
21   13C2-4:2FTS  18O2-PFHxS  ...                   7.4056  170.22
22    D3-NMeFOSA   13C4-PFOS  ...                   3.5214  171.44
23    13C8-PFOSA   13C4-PFOS  ...                   1.9558  174.74

[24 rows x 6 columns]

### 3.1.2 Concentration and recovery of EIS compounds 
*TO DO the weights of the bottles should be added as a column per sample*
*TO DO make a check if recovery falls in Table 5 + add RSD*


Here we compute the concentration of the EIS compounds based on the calculated response factor, the dilution factor and the sample volume (EPA 1633, page 44). 
It is important to enter the correct sample volume, otherwise calculations will be wrong. 

In [24]:
# computation of EIS concentration and recovery rate
from solutions import NIS_stock, EIS_stock
from allowed_recoveries import EIS_NIS_compound_recovery_aqueous
DF_sample = 1 #dilution factor 
mass_bottle_filled = 499.8 #grams
mass_bottle_empty = 53.68 #grams
Ws = (mass_bottle_filled - mass_bottle_empty)/1000 #L
Ws_cal = 0.25#ml


dilution_NIS_stock = 2 #1:1 dilution
dilution_NIS_stock_cal = 40 #dilution cal
dilution_EIS_stock_cal = 40 #dilution cal
spiked_amount_cal = 0.05 #ml
Df_cal = 1 #dilution factor
spiked_amount = 0.05 #ml

dilution_EIS_stock = 2 
spiked_amount_EIS =0.05 #ml

#calibration line
conc_EIS_cal = calculate_conc_EIS_cal(rawdata, eis_RFS, dilution_NIS_stock_cal, Ws_cal,
                          EIS_NIS_analogs, Df_cal, NIS_stock, spiked_amount_cal)

calculated_recoveries_EIS_cal = calculate_recoveries_EIS(conc_EIS_cal, EIS_stock, dilution_EIS_stock_cal, spiked_amount_cal, Ws_cal)

rse_results_Cal_EIS = calculate_eis_rse_modified(conc_EIS_cal, expected_concs_EIS, p=2)
validation_results_cal= validate_recoveries(calculated_recoveries_EIS_cal, EIS_NIS_compound_recovery_aqueous)
summary_cal = summarize_validation_results(validation_results_cal)
print("Validation Summary:")
for key, value in summary_cal['Overall'].items():
    print(f"{key}: {value}")

failed_recoveries_cal = get_failed_recoveries(validation_results_cal)



calculated_cons_EIS = calculate_conc_EIS(rawdata, eis_RFS, dilution_NIS_stock, Ws,EIS_NIS_analogs, DF_sample, NIS_stock, spiked_amount)

calculated_recoveries_EIS = calculate_recoveries_EIS(calculated_cons_EIS, EIS_stock, dilution_EIS_stock, spiked_amount_EIS, Ws)


conc_EIS_cal = calculate_conc_EIS_cal(rawdata, eis_RFS, dilution_NIS_stock_cal, Ws_cal,
                          EIS_NIS_analogs, Df_cal, NIS_stock, spiked_amount_cal)
# Run the validation
validation_results = validate_recoveries(calculated_recoveries_EIS, EIS_NIS_compound_recovery_aqueous)

# Get summary
summary = summarize_validation_results(validation_results)
print("Validation Summary:")
for key, value in summary['Overall'].items():
    print(f"{key}: {value}")

failed_recoveries = get_failed_recoveries(validation_results)

Name Type  ...  D7-NMeFOSE_Conc_Calc  13C4-PFBA_Conc_Calc
1  CS1  Cal  ...             24.989409            10.012923
2  CS2  Cal  ...             31.330097            10.049465
3  CS3  Cal  ...             21.875110             9.949198
4  CS4  Cal  ...             24.233892            10.007130
5  CS5  Cal  ...             24.176081             9.958111
6  CS6  Cal  ...             26.442630            10.019162
7  CS7  Cal  ...             21.941748            10.002228

[7 rows x 26 columns]

Name Type  ...  13C2-PFDoA_Recovery_%  13C2-PFTeDA_Recovery_%
1  CS1  Cal  ...              93.899140               90.115605
2  CS2  Cal  ...              92.300261               95.437576
3  CS3  Cal  ...             111.274273               96.452020
4  CS4  Cal  ...             109.622259              108.040887
5  CS5  Cal  ...              96.188774              100.679729
6  CS6  Cal  ...             100.623128              102.340431
7  CS7  Cal  ...              96.092165              106.933752

[7 rows x 26 columns]

EIS compounds found: 24
EIS compounds with expected concentrations: 24
Number of Cal samples found: 7

13C6-PFDA (p=2)
Point	Expected	Measured	Difference	(Diff)²/xi²	(Diff)²/xi²/(n-p)
--------------------------------------------------------------------------------
CAL 1	1.25		1.1391		-0.1109		0.007875	0.001575
CAL 2	1.25		1.2323		-0.0177		0.000200	0.000040
CAL 3	1.25		1.3558		0.1058		0.007158	0.001432
CAL 4	1.25		1.3234		0.0734		0.003447	0.000689
CAL 5	1.25		1.2054		-0.0446		0.001272	0.000254
CAL 6	1.25		1.2324		-0.0176		0.000198	0.000040
CAL 7	1.25		1.2614		0.0114		0.000083	0.000017
--------------------------------------------------------------------------------
Sum of relative squared errors: 0.004047
RSE = 100 * √(0.004047) = 6.361%

13C2-PFTeDA (p=2)
Point	Expected	Measured	Difference	(Diff)²/xi²	(Diff)²/xi²/(n-p)
--------------------------------------------------------------------------------
CAL 1	1.25		1.1264		-0.1236		0.009770	0.001954
CAL 2	1.25		1.1930		-0.0570		0.002082	0.00

Empty DataFrame
Columns: [Sample_Name, Sample_Type, Compound, Calculated_Recovery_%, Expected_Lower_%, Expected_Upper_%, Status, Compound_Type]
Index: []

Empty DataFrame
Columns: [Name, Type, 13C6-PFDA_Conc_Calc, 13C2-PFTeDA_Conc_Calc, 13C3-PFBS_Conc_Calc, 13C7-PFUnA_Conc_Calc, 13C9-PFNA_Conc_Calc, 13C3-PFHxS_Conc_Calc, 13C2-8:2FTS_Conc_Calc, 13C8-PFOS_Conc_Calc, 13C2-6:2FTS_Conc_Calc, D9-NEtFOSE_Conc_Calc, D3-NMeFOSA_Conc_Calc, D5-NEtFOSA_Conc_Calc, 13C2-4:2FTS_Conc_Calc, 13C4-PFHpA_Conc_Calc, D3-NMeFOSAA_Conc_Calc, 13C5-PFHxA_Conc_Calc, 13C8-PFOSA_Conc_Calc, 13C5-PFPeA_Conc_Calc, 13C2-PFDoA_Conc_Calc, 13C3-HFPO-DA_Conc_Calc, 13C8-PFOA_Conc_Calc, D5-NEtFOSAA_Conc_Calc, D7-NMeFOSE_Conc_Calc, 13C4-PFBA_Conc_Calc]
Index: []

Empty DataFrame
Columns: [Name, Type, D7-NMeFOSE_Recovery_%, D9-NEtFOSE_Recovery_%, 13C4-PFBA_Recovery_%, 13C3-HFPO-DA_Recovery_%, 13C5-PFPeA_Recovery_%, 13C2-4:2FTS_Recovery_%, 13C2-6:2FTS_Recovery_%, 13C2-8:2FTS_Recovery_%, D3-NMeFOSAA_Recovery_%, D5-NEtFOSAA_Recovery_%, 13C5-PFHxA_Recovery_%, 13C4-PFHpA_Recovery_%, 13C8-PFOA_Recovery_%, 13C3-PFBS_Recovery_%, 13C3-PFHxS_Recovery_%, 13C8-PFOS_Recovery_%, 13C8-PFOSA_Recovery_%, D3-NMeFOSA_Recovery_%, D5-NEtFOSA_Recovery_%, 13C9-PFNA_Recovery_%, 13C6-PFDA_Recovery_%, 13C7-PFUnA_Recovery_%, 13C2-PFDoA_Recovery_%, 13C2-PFTeDA_Recovery_%]
Index: []

Name Type  ...  D7-NMeFOSE_Conc_Calc  13C4-PFBA_Conc_Calc
1  CS1  Cal  ...             24.989409            10.012923
2  CS2  Cal  ...             31.330097            10.049465
3  CS3  Cal  ...             21.875110             9.949198
4  CS4  Cal  ...             24.233892            10.007130
5  CS5  Cal  ...             24.176081             9.958111
6  CS6  Cal  ...             26.442630            10.019162
7  CS7  Cal  ...             21.941748            10.002228

[7 rows x 26 columns]

KeyError: 'Status'

## 4 Quantitative determination of concentrations target analytes
In this section we will compute the concentration of all the analyzed PFAS compounds in the sample. Our current method is capable of analyzing 40 different compounds. Here, we distinguish between two categories to perform the calculation: 
- True Isotope Dilution Quantification (ID): 24 compounds are directly compared to their istopically labeled analog.
- Extracted Internal Standard Quantification (EIS): 16 compounds are quantified by comparing to a *different* isotopically labeled compound.

### 4.1 Response Ratio (RR) and Response Factor (RF) for target ID compounds and target EIS compounds.
Here we compute the response ratio for the ID compounds by using the formula from EPA 1633 on page 27 under section 10.3.3.2. For now we also fix some naming errors in the raw data for the target analytes. The only difference with the above calculation is that EIS analogs instead of direct pairs are used for the target EIS compounds. For ease, we call both RF.

- *Note 1*: NMeFoSA and NEtFoSa are calculated now but are not included in the PFAC30PAR ampoule we currently use in the lab, so these compounds cannot be in the calibration line. Use these results with care.
- *Note 2*:EPA 1633 mentions that PFTrDA recovery should improve by taking the average of the analog EIS compounds 13c2-PFTeDA and 13c2-PFDoa. Currently it only takes the EIS compound 13C2-PFTeDA as an analog.
- *Note 3*: EPA 1633 is capable of analyzing 40 compounds, our native standard solution only contains 30, so you will get errors that data for some compounds is not found. This is correct. If new ampoules are ordered for the other 10 compounds, everything needed for the calculation is there.
- *Note 4*: the FTS compounds only have 6 cal levels instead of 7.

In [7]:
from analogs_compounds import target_EIS_analogs

# Make a copy of rawdata to avoid modifying the original
rawdata_copy = rawdata.copy()
columns_to_rename = {}
for col in rawdata_copy.columns:
    new_col = col
    # Fix PFPHpA -> PFHpA
    if 'PFPHpA' in col:
        new_col = new_col.replace('PFPHpA', 'PFHpA')
    # Remove spaces from FTS compounds (e.g., "4:2 FTS" -> "4:2FTS")
    if ' FTS' in col:
        new_col = new_col.replace(' FTS', 'FTS')
    
    if new_col != col:
        columns_to_rename[col] = new_col

if columns_to_rename:
    rawdata_copy.rename(columns=columns_to_rename, inplace=True)

RF_target_analytes= calculate_average_RRs_targets(rawdata_copy, expected_concs_EIS, calibration_solutions, target_EIS_analogs)

Found 40 total target-IS pairs:
  NMeFOSE -> D7-NMeFOSE
  NEtFOSE -> D9-NEtFOSE
  PFBA -> 13C4-PFBA
  HFPO-DA -> 13C3-HFPO-DA
  PFPeA -> 13C5-PFPeA
  4:2FTS -> 13C2-4:2FTS
  6:2FTS -> 13C2-6:2FTS
  8:2FTS -> 13C2-8:2FTS
  NMeFOSAA -> D3-NMeFOSAA
  NEtFOSAA -> D5-NEtFOSAA
  PFHxA -> 13C5-PFHxA
  PFHpA -> 13C4-PFHpA
  PFOA -> 13C8-PFOA
  PFBS -> 13C3-PFBS
  PFHxS -> 13C3-PFHxS
  PFOS -> 13C8-PFOS
  PFOSA -> 13C8-PFOSA
  NMeFOSA -> D3-NMeFOSA
  NEtFOSA -> D5-NEtFOSA
  PFNA -> 13C9-PFNA
  PFDA -> 13C6-PFDA
  PFUnA -> 13C7-PFUnA
  PFDoA -> 13C2-PFDoA
  PFTeDA -> 13C2-PFTeDA
  PFTrDA -> 13C2-PFTeDA
  PFPeS -> 13C3-PFHxS
  PFHpS -> 13C8-PFOS
  PFNS -> 13C8-PFOS
  PFDS -> 13C8-PFOS
  PFDoS -> 13C8-PFOS
  ADONA -> 13C3-HFPO-DA
  PFMPA -> 13C5-PFPeA
  PFMBA -> 13C5-PFPeA
  NFDHA -> 13C5-PFHxA
  9Cl-PF3ONS -> 13C3-HFPO-DA
  11Cl-PF3OUdS -> 13C3-HFPO-DA
  PFEESA -> 13C5-PFHxA
  3:3FTCA -> 13C5-PFPeA
  5:3FTCA -> 13C5-PFHxA
  7:3FTCA -> 13C5-PFHxA


AVERAGE RESPONSE RATIOS FOR TARGET COMPOUNDS WITH

Target_Compound   EIS_Isotope  ...  N_Levels                 Levels_Used
0          NMeFOSE    D7-NMeFOSE  ...         7  L1, L2, L3, L4, L5, L6, L7
1          NEtFOSE    D9-NEtFOSE  ...         7  L1, L2, L3, L4, L5, L6, L7
2             PFBA     13C4-PFBA  ...         7  L1, L2, L3, L4, L5, L6, L7
3          HFPO-DA  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7
4            PFPeA    13C5-PFPeA  ...         7  L1, L2, L3, L4, L5, L6, L7
5           4:2FTS   13C2-4:2FTS  ...         6      L1, L2, L3, L4, L5, L6
6           6:2FTS   13C2-6:2FTS  ...         6      L1, L2, L3, L4, L5, L6
7           8:2FTS   13C2-8:2FTS  ...         6      L1, L2, L3, L4, L5, L6
8         NMeFOSAA   D3-NMeFOSAA  ...         7  L1, L2, L3, L4, L5, L6, L7
9         NEtFOSAA   D5-NEtFOSAA  ...         7  L1, L2, L3, L4, L5, L6, L7
10           PFHxA    13C5-PFHxA  ...         7  L1, L2, L3, L4, L5, L6, L7
11           PFHpA    13C4-PFHpA  ...         7  L1, L2, L3, L4, L5, L6, L7
12            PFOA     13C8-PFOA  ...         7  L1, L2, L3, L4, L5, L6, L7
13            PFBS     13C3-PFBS  ...         7  L1, L2, L3, L4, L5, L6, L7
14           PFHxS    13C3-PFHxS  ...         7  L1, L2, L3, L4, L5, L6, L7
15            PFOS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
16           PFOSA    13C8-PFOSA  ...         7  L1, L2, L3, L4, L5, L6, L7
17         NMeFOSA    D3-NMeFOSA  ...         7  L1, L2, L3, L4, L5, L6, L7
18         NEtFOSA    D5-NEtFOSA  ...         7  L1, L2, L3, L4, L5, L6, L7
19            PFNA     13C9-PFNA  ...         7  L1, L2, L3, L4, L5, L6, L7
20            PFDA     13C6-PFDA  ...         7  L1, L2, L3, L4, L5, L6, L7
21           PFUnA    13C7-PFUnA  ...         7  L1, L2, L3, L4, L5, L6, L7
22           PFDoA    13C2-PFDoA  ...         7  L1, L2, L3, L4, L5, L6, L7
23          PFTeDA   13C2-PFTeDA  ...         7  L1, L2, L3, L4, L5, L6, L7
24          PFTrDA   13C2-PFTeDA  ...         7  L1, L2, L3, L4, L5, L6, L7
25           PFPeS    13C3-PFHxS  ...         7  L1, L2, L3, L4, L5, L6, L7
26           PFHpS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
27            PFNS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
28            PFDS     13C8-PFOS  ...         7  L1, L2, L3, L4, L5, L6, L7
29           ADONA  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7
30      9Cl-PF3ONS  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7
31    11Cl-PF3OUdS  13C3-HFPO-DA  ...         7  L1, L2, L3, L4, L5, L6, L7

[32 rows x 8 columns]

### 4.2 Calculating Concentrations target analytes
Here we compute the concentration of the target analytes (ng/L) based on the formula on page 44 of EPA method 1633. 

*Note 1* % solids is not added yet, needs to be done before the solids are added. 


In [31]:

conc_target_analytes = calculate_conc_targets(rawdata_copy,
                           RF_target_analytes,
                           EIS_stock, dilution_EIS_stock, spiked_amount_EIS, Ws,
                           DF_sample,  debug_compound=None )

Name    Type  ...  9Cl-PF3ONS_Conc_Calc  11Cl-PF3OUdS_Conc_Calc
0           blank + EIS  Sample  ...              0.000000                0.012141
1                QC std  Sample  ...            111.748518              113.542192
2                   KCS  Sample  ...             97.586505               94.680232
3                   DWZ  Sample  ...              0.001545                0.008544
4                   KWZ  Sample  ...              0.004931                0.009088
5                   DWP  Sample  ...            115.605196               95.840190
6                   KWP  Sample  ...             95.516111               78.988097
7                   KCS  Sample  ...             92.789094               90.604086
8                 QCstd  Sample  ...            109.461237              108.304356
9                 rinse  Sample  ...             31.024046               57.179464
10                rinse  Sample  ...            124.096183               57.179464
11                rinse  Sample  ...             41.365394               57.179464
12  blank_spike_eis-nis  Sample  ...              0.000000                0.004781
13                1_ccf  Sample  ...              0.006463                0.000000
14                 2_cf  Sample  ...              0.003284                0.012107
15                  3_f  Sample  ...              0.008029                0.004933
16               4_none  Sample  ...              0.002839                0.005233
17                rinse  Sample  ...                   inf                     NaN
25                RINSE  Sample  ...             31.024046               28.589732
26          blank_spike  Sample  ...              0.004970                0.004580
27             alex_kwz  Sample  ...              0.002280                0.008403
28             alex_dwz  Sample  ...              0.003181                0.005862
29          alex_kwpfas  Sample  ...            105.787820               92.530578
30          alex_dwpfas  Sample  ...            112.487890              105.742909

[24 rows x 34 columns]

## 5. Save to Excel

In [32]:

# Define the Excel file path
excel_file_path = "results/EPA_1633_results.xlsx"
# Save DataFrames to separate sheets
with pd.ExcelWriter(excel_file_path, engine='openpyxl') as writer:
    # 2.1 RSE EIS compounds
    df_rse_eis.to_excel(writer, sheet_name="2.1_RSE_EIS_Compounds", index=False)
    # 2.2 RSE target analytes
    df_rse_target.to_excel(writer, sheet_name="2.2_RSE_target", index=False)
    # 3.1 Concentration EIS compounds
    eis_RFS.to_excel(writer, sheet_name="3.1.1_RF_Eis", index=False)
    calculated_cons_EIS.to_excel(writer, sheet_name="3.1.2_concentration_EIS", index=False)

    #3.2 Recoveries EIS compounds
    calculated_recoveries_EIS.to_excel(writer, sheet_name="3.2_recoveries_EIS", index=False)

    #4.1 Target analytes response factors 
    RF_target_analytes.to_excel(writer, sheet_name="4.1 Target_analytes_RF", index=False)
    # 4.2 Target analytes concentrations
    conc_target_analytes.to_excel(writer, sheet_name="4.2_Target_Conc", index=False)

print(f"DataFrames saved to {excel_file_path}")


DataFrames saved to results/EPA_1633_results.xlsx
